### DISCUSSION TITLE
How to safely copy json objects without memory leak?

### DISCUSSION THREAD

**Discussion by zliucd** [nlohmann/json#3764] [General]
Hi,

I would like to pass ```json``` objects in two classes, and memory leaks were found by Address Sanitizer. How to safely copy/pass json objects?  I'm looking for a safe and also memory-efficient solution (without extra copy, if possible). Thanks a lot.
 
```
class A
{
   json j;
}

class B
{
  std::vector<json> vec_j;
}

void foo()
{

    A *a = new A();
    // do something, a->j  has some value;
   
   B *b = new B();
   b->vec_j.push_back(a->j);     // only one item for demonstration; there could be multiple items.

  delete(a);         // AddressSanitizer reported memory leak.  'a' must be deleted before deleting 'b'.
                          
  delete(b);
}

```

Update a working example **which does not trigger ASAN leak reports.**
```
#include <vector>
#include "json.hpp"

using namespace  std;
using ordered_json = nlohmann::ordered_json;

class A
{
public:
    ordered_json j;
};

class B
{
public:
    vector<ordered_json> vec_j;
};

int main()
{
    A *a = new A();
    B *b = new B();

    a->j["name"] = "orange";
    b->vec_j.push_back(a->j);

    delete(a);
    delete(b);

}

```

**Comment by nlohmann** [comment]
Please share a working example and the actual output of ASAN.

**Reply by zliucd** [reply]
I've updated a working example, which does not trigger leak report.  You may close this issue as a "false positive".

I'm not familiar with C++ copying; when using `vector.push_back(json j)`, another copy of `j` is actually created and pushed  in the vector?

**Reply by falbrechtskirchinger** [reply]
> I'm not familiar with C++ copying; when using `vector.push_back(json j)`, another copy of `j` is actually created and pushed in the vector?

Yes. You can move `j` using `vector.push_back(std::move(j))`, after which `j` won't hold a value (and the variable has to be re-initialized if you want to continue using it).

Also: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete

**Reply by zliucd** [reply]
Thanks, ```std::move()``` is pretty useful without allocating another memory piece.

**Comment by falbrechtskirchinger** [comment]
The code provided (with syntax errors, etc. fixed) does not trigger any ASAN errors.
